src/Storages/MergeTree/MergeTreeIndexTextPostingListCodec.cpp

```diff
@@ -0,0 +1,358 @@
+#include <Storages/MergeTree/MergeTreeIndexTextPostingListCodec.h>
+#include <Storages/MergeTree/MergeTreeIndexText.h>
+#include <Storages/MergeTree/IntegerCodecTrait.h>
+
+namespace DB
+{
+/// PostingListCodecImpl
+///
+/// A serializer/deserializer for a postings list (sorted docIDs) stored in a compact
+/// block-compressed format.
+///
+/// High-level idea:
+/// - Input docIDs are strictly increasing.
+/// - Values are encoded as deltas (gaps) and compressed in fixed-size blocks (kBlockSize).
+/// - Each compressed block is stored as: [1 byte bits-width][bitpacked payload].
+/// - A posting list may be split into "segments" (logical chunks, controlled by postings_list_block_size)
+///   to simplify metadata and to support multiple ranges per token (min/max docID per segment).
+///
+class PostingListCodecImpl
+{
+    /// Per-segment header written before each segment payload.
+    ///
+    /// - bytes: number of compressed bytes following this header (segment payload size)
+    /// - cardinality: number of postings (docIDs) in this segment
+    /// - base_value: the first docID in the segment (also used to restore deltas)
+    struct ContainerHeader
+    {
+        ContainerHeader() = default;
+        explicit ContainerHeader(size_t bytes_, uint32_t cardinality_, uint32_t base_value_)
+            : bytes(bytes_)
+            , cardinality(cardinality_)
+            , base_value(base_value_)
+        {
+        }
+
+        template<typename Out>
+        void encodeTo(Out & out) const
+        {
+            writeVarUInt(bytes, out);
+            writeVarUInt(cardinality, out);
+            writeVarUInt(base_value, out);
+        }
+
+        template<typename In>
+        void decodeFrom(In & in)
```

Comment:
Re l. 37/45, can we name them `serialize` and `deserialize` or `read`/`write`?

Reply:
Ok, `read/write` is better. I have fixed it.